File: pulpo/graphs/nodes/metric.py
"""
MetricNode — registra una métrica de negocio y, opcionalmente, notifica a un
sistema externo interesado vía webhook.

Guardar la métrica es el paso crítico. El webhook es un side-effect
best-effort dirigido a un sistema externo (ej: uno que integra Pulpo y quiere
enterarse en tiempo real de qué piden los usuarios): si falla, NO aborta el
flow — pero el fallo queda logueado a nivel ERROR para poder detectarlo.
"""
import json
import logging

from .base import BaseNode, interpolate
from .state import FlowState

logger = logging.getLogger(__name__)


class MetricNode(BaseNode):
    async def run(self, state: FlowState) -> FlowState:
        metric_name = interpolate(self.config.get("metric_name", ""), state).strip()
        if not metric_name:
            logger.warning("[MetricNode] metric_name vacío — no se registra nada")
            return state

        value = interpolate(str(self.config.get("value", "")), state)

        raw_metadata = self.config.get("metadata") or {}
        metadata = {k: interpolate(str(v), state) for k, v in raw_metadata.items()} if isinstance(raw_metadata, dict) else {}

        from pulpo.core import db
        await db.insert_metric(
            bot_id=state.bot_id or "",
            contact_phone=state.contact_phone or "",
            contact_name=state.contact_name or "",
            canal=state.canal or "",
            metric_name=metric_name,
            value=value,
            metadata=json.dumps(metadata, default=str) if metadata else None,
        )
        logger.info("[MetricNode] %s=%s (bot=%s, contact=%s)", metric_name, value, state.bot_id, state.contact_phone)

        webhook_url = (self.config.get("webhook_url") or "").strip()
        if webhook_url:
            await self._notify_webhook(webhook_url, {
                "metric_name":   metric_name,
                "value":         value,
                "bot_id":        state.bot_id,
                "contact_phone": state.contact_phone,
                "contact_name":  state.contact_name,
                "canal":         state.canal,
                "metadata":      metadata,
            })

        return state

    @staticmethod
    async def _notify_webhook(url: str, payload: dict) -> None:
        """Fire-and-forget: un fallo no interrumpe el flow, pero queda en el log
        (nivel ERROR) para poder detectarlo. No hay reintentos."""
        try:
            import httpx
            async with httpx.AsyncClient(timeout=10) as client:
                resp = await client.post(url, json=payload)
                resp.raise_for_status()
        except Exception as e:
            logger.error("[MetricNode] webhook falló url=%s metric=%s: %s", url, payload.get("metric_name"), e)

    @classmethod
    def config_schema(cls) -> dict:
        return {
            "metric_name": {
                "type":     "string",
                "label":    "Nombre de la métrica",
                "default":  "",
                "hint":     "Ej: intencion_detectada, producto_consultado. Soporta {{templates}}",
                "required": True,
            },
            "value": {
                "type":     "string",
                "label":    "Valor",
                "default":  "",
                "hint":     "Soporta {{templates}}, ej: {{message}} o {{route}}",
                "required": True,
            },
            "metadata": {
                "type":    "json",
                "label":   "Metadata extra (opcional)",
                "default": {},
                "hint":    '{"campo": "{{template}}"}',
            },
            "webhook_url": {
                "type":    "string",
                "label":   "Webhook (opcional)",
                "default": "",
                "hint":    "POST con {metric_name, value, bot_id, contact_phone, contact_name, canal, metadata}. "
                           "Fire-and-forget: si falla no interrumpe el flow, el error queda logueado.",
            },
        }

File: web/lib/nodes/metric.ts
import { getDb } from "@/lib/db/client";
import { metrics } from "@/lib/db/schema";
import type { NodeDef } from "./base";
import { interpolate } from "./interpolate";

// TS port of MetricNode (pulpo/graphs/nodes/metric.py). Saving the metric is
// the critical step; the webhook notification is a best-effort side-effect
// (logged on failure, never aborts the flow) -- same contract as the Python
// original.
export const metricNode: NodeDef = {
  label: "Métrica",
  color: "#a16207",
  description: "Registra una métrica de negocio en DB y, opcionalmente, notifica a un sistema externo vía webhook.",
  configSchema: {},
  async run(state, config) {
    const metricName = interpolate((config.metric_name as string) ?? "", state).trim();
    if (!metricName) {
      console.warn("[metric] metric_name vacío — no se registra nada");
      return state;
    }

    const value = interpolate(String(config.value ?? ""), state);
    const rawMetadata = (config.metadata as Record<string, unknown>) ?? {};
    const metadata: Record<string, string> = {};
    if (rawMetadata && typeof rawMetadata === "object") {
      for (const [k, v] of Object.entries(rawMetadata)) {
        metadata[k] = interpolate(String(v), state);
      }
    }

    await getDb().insert(metrics).values({
      botId: state.botId || "",
      contactPhone: state.contactPhone || "",
      contactName: state.contactName || "",
      canal: state.canal || "",
      metricName,
      value,
      metadata: Object.keys(metadata).length ? JSON.stringify(metadata) : null,
    });

    const webhookUrl = ((config.webhook_url as string) ?? "").trim();
    if (webhookUrl) {
      try {
        const res = await fetch(webhookUrl, {
          method: "POST",
          headers: { "Content-Type": "application/json" },
          body: JSON.stringify({
            metric_name: metricName,
            value,
            bot_id: state.botId,
            contact_phone: state.contactPhone,
            contact_name: state.contactName,
            canal: state.canal,
            metadata,
          }),
        });
        if (!res.ok) throw new Error(`HTTP ${res.status}`);
      } catch (err) {
        console.error(`[metric] webhook falló url=${webhookUrl} metric=${metricName}`, err);
      }
    }

    return state;
  },
};
